### PR TITLE
feat: add shuttle golang actions as an execution type

### PR DESCRIPTION
**Pull request by kjuulh**
This is still a bit wip, as arguments aren't currently supported all the way through shuttle, but as a first step this enables shuttle to build and run simple shuttletask/*.go commands.

Do note: the change doesn't contain as many files as it may seem, a lot of them is just testdata/test projects

### CODE REVIEW

**Review by mahlunar — COMMENTED** [review]
Very nice. I think we also need some documentation to go along with this.

File: examples/moon-base/shuttletask/go.mod [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module shuttletask
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply examples/moon-base/shuttletask/go.mod]: Renamed it to actions for now.

File: pkg/executors/golang/cmder/cmder.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+		rawJson, err := json.Marshal(cmdNames)
+		if err != nil {
+			return err
+		}
+
+		_, err = fmt.Printf("%s", string(rawJson))
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/cmder/cmder.go:53]: Added a comment

File: pkg/executors/golang/cmder/cmder.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+		rootcmd.AddCommand(cobracmd)
+	}
+
+	if err := rootcmd.Execute(); err != nil {
+		log.Fatalf("%v", err)
+	}
+}
+
+type Arg struct {
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/cmder/cmder.go:94]: Only if they need to be used by third party libraries at some point. For now lets private them

[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/cmder/cmder.go:94]: Correction they are used in codegen, hence the naming. 

I.e. it is the structs and functions used in the temporary main.go file. See codegen for specifics.

File: pkg/executors/golang/codegen/compilation.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	"path"
+)
+
+func CompileBinary(ctx context.Context, shuttlelocaldir string) (string, error) {
+	cmd := exec.Command("go", "build")
+	cmd.Dir = path.Join(shuttlelocaldir, "tmp")
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Printf("%s\n", string(output))
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/codegen/compilation.go:16]: changed to use UI instead. this should hide it unless needed with Verbose

File: pkg/executors/golang/codegen/format.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	"path"
+)
+
+func Format(ctx context.Context, shuttlelocaldir string) error {
+	cmd := exec.Command("go", "fmt", "./...")
+	cmd.Dir = path.Join(shuttlelocaldir, "tmp")
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Printf("%s\nerror: %v", string(output), err)
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/codegen/format.go:16]: changed to use UI instead. this should hide it unless needed with Verbose

File: pkg/executors/golang/codegen/tidy.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+	"path"
+)
+
+func ModTidy(ctx context.Context, shuttlelocaldir string) error {
+	cmd := exec.Command("go", "mod", "tidy")
+	cmd.Dir = path.Join(shuttlelocaldir, "tmp")
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Printf("%s\n", string(output))
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/codegen/tidy.go:16]: changed to use UI instead. this should hide it unless needed with Verbose

File: pkg/executors/golang/compile/matcher/matcher.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+func BinaryMatches(
+	ctx context.Context,
+	hash string,
+	shuttletask *discover.ShuttleTaskDiscovered,
+) (string, bool, error) {
+	shuttlebindir := path.Join(shuttletask.ParentDir, ".shuttle/shuttletask/binaries")
+
+	if _, err := os.Stat(shuttlebindir); errors.Is(err, os.ErrNotExist) {
+		log.Println("DEBUG: package doesn't exist continueing")
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/compile/matcher/matcher.go:28]: Changed to UI

File: pkg/executors/golang/compile/compile.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		return "", err
+	}
+
+	binaryPath, ok, err := matcher.BinaryMatches(ctx, hash, shuttletask)
+	if err != nil {
+		return "", err
+	}
+
+	if ok && !alwaysBuild {
+		log.Printf("DEBUG: file already matches continueing\n")
```
[the remark this answers is not in the document]
Comment (kjuulh) [reply pkg/executors/golang/compile/compile.go]: Changed to UI

File: pkg/executors/golang/executer/run.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+	"log"
+
+	"github.com/lunarway/shuttle/pkg/config"
+	"github.com/lunarway/shuttle/pkg/ui"
+)
+
+func Run(ctx context.Context, ui *ui.UI, c *config.ShuttleProjectContext, path string, args ...string) error {
+	binaries, err := prepare(ctx, ui, path, c)
+	if err != nil {
+		log.Printf("failed to run command: %v", err)
```
Comment (mahlunar) [inline pkg/executors/golang/executer/run.go:14]: Remove?

File: pkg/executors/golang/executer/run_test.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func updateShuttle(t *testing.T, path string) {
+	err := os.RemoveAll("testdata/child/.shuttle/")
+	assert.NoError(t, err)
+
+	shuttleCmd := exec.Command("shuttle", "ls")
+	shuttleCmd.Dir = path
+	if output, err := shuttleCmd.CombinedOutput(); err != nil {
+		log.Printf("%s\n", string(output))
```
Comment (mahlunar) [inline pkg/executors/golang/executer/run_test.go:41]: remove?

**Review by mahlunar — COMMENTED** [review]
I think you should check each exposed type in the pkg folders and consider if should be kept internal, Otherwise I think we are close to get this merged 🚀

File: pkg/executors/golang/executer/run_test.go (hunk 10 above)
Comment (kjuulh) [reply pkg/executors/golang/executer/run_test.go:41]: We will keep this as it is running only for a test function.

**Comment by kjuulh** [comment]
@mahlunar The exported types are fine for now. We should consider moving some of the shuttle stuff to internal instead. A lot of the types needs to be exported because they are used by other internal packages, not because they are used externally.

**Comment by mahlunar** [comment]
> @mahlunar The exported types are fine for now. We should consider moving some of the shuttle stuff to internal instead. A lot of the types needs to be exported because they are used by other internal packages, not because they are used externally.

I think that is a good idea. Do you want to do that now?

**Comment by kjuulh** [comment]
@mahlunar Nope, I will do it in a separate pr to not clutter it too much. It may also require some refactoring.

**Review by mahlunar — APPROVED** [review]

**Pull request merged**